README.md
# Twitch Plays!
[![wakatime](https://wakatime.com/badge/github/veryCrunchy/TwitchPlays.svg?style=flat)](https://wakatime.com/@veryCrunchy/projects/xyomfgjwhr)
###### [CHANGELOG](https://github.com/veryCrunchy/TwitchPlays/blob/main/CHANGELOG.md) | [TRELLO](https://trello.com/b/zx5duUS3/bigmac)

### Dependencies

- [Node.js](https://nodejs.org/)

### Installing

- Open a terminal and clone this repo using [git](https://git-scm.com/)

```
git clone https://github.com/veryCrunchy/TwitchPlays.git
```

- Open a terminal in the folder

```
cd TwitchPlays
```

or

```
cd C:/Path/To/TwitchPlays
```

- Run the following command (make sure [Node.js](https://nodejs.org/) is installed)

```
npm i
```

- create a file named `.env` and copy/paste the contents from `.env.example` into it
- modify accordingly

### Executing program

- Open a terminal in the folder of this program
- Run the following command

```
nodemon .
```

## Author

[@veryCrunchy](https://github.com/veryCrunchy)

## License

Before (commercially) using this program, contact the author via Discord or Email

- veryCrunchy#3281
- verycrunchydev@gmail.com

## Acknowledgments

Client

- [GreasyMac](https://greasygang.co)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [Unreleased]

### Added

- [Website]: Better chat inputs overlay
- [Website]: Controls overlay
- Game Inputs: Chances on chaos mode
- [Input Preset](https://github.com/veryCrunchy/TwitchPlays/blob/main/preset/Peggle_Deluxe.json): Peggle Deluxe
<!-- - Force stop command for mods and broadcaster; tp!stop -->

## [1.1.1] - 2023-06-22

### Added

- [Website]​: api
- [Website]: View current inputs.
- [Input Preset](https://github.com/veryCrunchy/TwitchPlays/blob/main/preset/Mario_Kart_Wii-RMCE01.json): Mario Kart Wii
- Configuration: Set the alignment of the obs overlay (left, center, right)

### Fixed

- Releasing key too early when two inputs are pressing the same key simultaneously
- Not using delay when time is set bellow 0.3

## [1.1.0] - 2023-06-11

### Added

- Started keeping a changelog
- Obs overlay
- Game inputs: Delay support
- Game inputs: Mouse movement support
- New setting for TIMED mode; `TIMED_ON_INPUT`
- `.env` configurations to game input configurations, allows changing settings per game
- Timeout for CHAOS mode

### Changed

- Automatic Input Switching: Improved accuracy
- Automatic Input Switching: Non wii game support
- Configuration: Per game configuration
- Configuration: Configurations are now set by default, you can modify the defaults in `.env`
  Will be overwritten by per game configurations
- Game inputs: Support for multiple/combined outputs instead of just one
- Game inputs: Support for multiple times to go with multiple inputs

### Fixed

- "Not focused..." spam; Now only shows once until focused again.

### Removed

- Wii game only restriction

## [1.0.0] - 2023-06-05

### Added

- Support for multiple game inputs using `inputs/<game>.json`
- Input preset: M&M Kart Racing

### Removed

- Input configuration inside of index.js

### Fixed

- TIMED Mode: One input per user per interval

[website]: https://twitchplays.greasygang.co
[unreleased]: https://github.com/veryCrunchy/TwitchPlays/compare/v1.1.1...HEAD
[1.1.1]: https://github.com/veryCrunchy/TwitchPlays/compare/v1.1.0...v1.1.1
[1.1.0]: https://github.com/veryCrunchy/TwitchPlays/compare/v1.0.0...v1.1.0
[1.0.0]: https://github.com/veryCrunchy/TwitchPlays/releases/tag/v1.0.0


configs.js
//Change all the configurations inside of your .env file
//This file is just for guidance of what the configuration do
//Do not change anything in this file!!

require("dotenv").config();
let configuration = {
  // Your twitch channel name, as shown in your twitch.tv/channel_name url. Should be lowercase!
  CHANNEL: process.env.CHANNEL,

  ///////////////////////////////////////////////////////////////
  // ↓ ↓ ↓ ↓ ↓ ↓ ↓ ↓ ↓ ↓ CONFIG EXPLANATION ↓ ↓ ↓ ↓ ↓ ↓ ↓ ↓ ↓ //
  /////////////////////////////////////////////////////////////
  // Will be overwritten by individual game configurations! //
  ///////////////////////////////////////////////////////////

  // Timed mode will output the most inputted action every INTERVAL seconds.
  // Chaos mode will output all inputs, waits TIMEOUT seconds before the same user is able to input again.
  TIMED_MODE: true, // true = timed mode, false = chaos mode || Will be overwritten by inputs file

  ///// TIMED MODE \\\\\

  // true = will only start interval timer when someone sends a message || RECOMMENDED
  // false = Will continuously run on interval, sometimes allowing chatters to sneak in their own input, without being affected by other chatters.
  //TIMED_ON_INPUT = "true"
  TIMED_ON_INPUT: true,

  //INTERVAL = "2"
  INTERVAL: 2,

  ///// CHAOS MODE \\\\\
  // TIMEOUT = "2"
  TIMEOUT: 2,

  ///// OTHER \\\\\
  // WINDOW = ["Dolphin", "RMCE01"]
  WINDOW: ["Dolphin"], // Window name must include all these words for the inputs to register, to prevent chat from taking over your computer.
};

console.log();

if (process.env.TIMED_MODE != undefined)
  configuration.TIMED_MODE = process.env.TIMED_MODE?.toLowerCase() === "true";

if (process.env.TIMED_ON_INPUT != undefined)
  configuration.TIMED_ON_INPUT =
    process.env.TIMED_ON_INPUT?.toLowerCase() === "true";

if (process.env.INTERVAL != undefined)
  configuration.INTERVAL = Number(process.env.INTERVAL);

if (process.env.TIMEOUT != undefined)
  configuration.TIMEOUT = Number(process.env.TIMEOUT);

if (process.env.WINDOW != undefined)
  configuration.WINDOW = JSON.parse(process.env.WINDOW);

if (process.env.OBS != undefined) configuration.OBS = String(process.env.OBS);

module.exports = {
  getConfiguration: () => configuration,
  updateConfiguration: (updatedValues) => {
    configuration = { ...configuration, ...updatedValues };
  },
};
require("./src/index.js");


src/func/go.js
const {
  Key,
  Point,
  Button,
  keyboard,
  mouse,
  screen,
  left,
  right,
  up,
  down,
  straightTo,
} = require("@nut-tree/nut-js");
const { findBestMatch } = require("string-similarity");
const { getConfiguration } = require("../../configs.js");
let holding = new Map();
async function hold(keys, seconds) {
  keyboard.pressKey(keys);
  let hold = holding.get(keys);
  hold == undefined ? holding.set(keys, 1) : holding.set(keys, hold + 1);
  setTimeout(async function () {
    hold = holding.get(keys);
    if (hold === 1) {
      keyboard.releaseKey(keys);
      holding.set(keys, hold - 1);
    } else holding.set(keys, hold - 1);
  }, seconds * 1000);
}
function getMatch(m) {
  if (typeof m != "string" || m == null)
    return { output: undefined, time: undefined, name: undefined };
  m = m.trim().toLowerCase();
  const data = getConfiguration().data;

  let match = findBestMatch(m, Array.from(data.inputs.keys()));
  if (match) {
    if (match.bestMatch.rating < 0.5)
      return { output: undefined, time: undefined, name: undefined };
    match = match.bestMatch.target;
  } else return;

  const name = data.inputs.get(match);
  const time = Number(data.times.get(name)) || 1;
  const output = data.outputs.get(name);

  return { output: output, time: time, name: name };
}
mouse.config.mouseSpeed = 2000;

async function manage(name) {
  const env = getConfiguration();
  const times = env.data.times.get(name);
  const delays = env.data.delays.get(name);

  for (let [i, press] of env.data.outputs.get(name).entries()) {
    let delay = 0;
    if (delays) {
      delay = delays[i];
    }
    switch (true) {
      case press.startsWith("mouse(center)"):
        moveMouseToCenter();
        break;
      case press.startsWith("left"): {
        let m = Number(press.match(/left\((\d+)\)/)[1]);
        goMouse(left(m), delay);
        break;
      }
      case press.startsWith("right"): {
        let m = Number(press.match(/right\((\d+)\)/)[1]);
        goMouse(right(m), delay);
        break;
      }
      case press.startsWith("up"): {
        let m = Number(press.match(/up\((\d+)\)/)[1]);
        goMouse(up(m), delay);
        break;
      }
      case press.startsWith("down"): {
        let m = Number(press.match(/down\((\d+)\)/)[1]);
        goMouse(down(m), delay);
        break;
      }
      case press.startsWith("click(left)"): {
        await mouse.click(Button.LEFT);
        break;
      }
      case press.startsWith("click(right)"): {
        await mouse.click(Button.RIGHT);
        break;
      }
      case press.startsWith("type"): {
        let m = press.match(/type\((.+)\)/)[1];
        if (delay) {
          setTimeout(async () => {
            keyboard.type(m);
          }, delay * 1000);
        } else {
          keyboard.type(m);
        }
        break;
      }
      default: {
        const time = times[i];
        if (delay) {
          if (time == 0) {
            setTimeout(() => {
              keyboard.pressKey(Key[press]);
              setTimeout(() => keyboard.releaseKey(Key[press]), 70);
            }, delay * 1000);
          } else {
            setTimeout(() => {
              hold(Key[press], time);
            }, delay * 1000);
          }
        } else if (time == 0) {
          keyboard.pressKey(Key[press]);
          setTimeout(() => keyboard.releaseKey(Key[press]), 70);
        } else {
          hold(Key[press], time);
        }
        break;
      }
    }
  }
}

async function goMouse(func, m, delay) {
  if (delay) {
    setTimeout(async () => {
      mouse.move(func);
    }, delay * 1000);
  } else {
    mouse.move(func);
  }
}

async function mouseImage(path) {
  await mouse.move(
    straightTo(centerOf(screen.find(imageResource(`../../img/${path}`))))
  );
}

async function moveMouseToCenter() {
  const centerX = Math.round((await screen.width()) / 2);
  const centerY = Math.round((await screen.height()) / 2);
  console.log(screen.height());
  let m = await mouse.move(straightTo(new Point(centerX, centerY)));
  console.log(m);
}

module.exports = { hold, getMatch, manage };


src/func/obs.js
const { getConfiguration } = require("../../configs.js");
const express = require("express");
const app = express();
const http = require("http");
const port = 4000;
const server = http.createServer(app);
const { Server } = require("socket.io");
const io = new Server(server);
let connected = false;
io.on("connection", (socket) => {
  socket.on("disconnect", () => {
    if (connected)
      console.log(
        `\x1b[33m -\x1b[0m\x1b[2m\x1b[31m Obs overlay disconnected\x1b[0m`
      ),
        (connected = false);
  });
  if (!connected)
    console.log(
      `\x1b[33m -\x1b[0m\x1b[2m\x1b[36m Obs overlay connected\x1b[0m`
    ),
      (connected = true);
  obs.clear();
  const env = getConfiguration();
  io.emit("align", env.OBS);
  io.emit("connection");
});

server.listen(port, () => {
  console.log(`\x1b[33m -\x1b[0m\x1b[32m\x1b[2m Started obs overlay\x1b[0m`);
});

const obs = {
  send: function (msg) {
    io.emit("message", msg);
  },
  clear: function () {
    io.emit("clear");
  },
  emit: function (event, msg) {
    io.emit(event, msg);
  },
};

app.get("/", (req, res) => {
  res.sendFile(__dirname + "/index.html");
});

module.exports.obs = obs;


src/func/chaos.js
const { go, obs, getConfiguration } = require("./");
const { Key, getActiveWindow } = require("@nut-tree/nut-js");
let tempUser = new Map();
let preserved = false;
async function message(tags, m, self) {
  let user = tempUser.get(tags.username);
  if (user) return;
  const env = getConfiguration();

  const windowRef = await getActiveWindow();
  const [title] = await Promise.all([windowRef.title]);
  const keywords = env.WINDOW;
  if (
    !keywords
      .map((keyword) => keyword.toLowerCase())
      .every((keyword) => title.toLowerCase().includes(keyword))
  ) {
    if (!preserved)
      console.log(
        "\x1b[2m\x1b[33mNot focused on correct window, cancelled input.\x1b[0m"
      ),
        obs.send(
          `<span class="yellow edim">Not focused on correct window, cancelled input.</span>`
        ),
        (preserved = true);
    return;
  }
  preserved = false;

  if (env.TIMEOUT !== 0) {
    tempUser.set(tags.username, true);
    setTimeout(function () {
      tempUser.delete(tags.username);
    }, env.TIMEOUT * 1000);
  }

  let { name } = go.getMatch(m);
  if (!name) return;
  let chance = env.data.chances.get(name) || undefined;
  if (chance != undefined && Math.random() * 100 > env.data.chances.get(name)) {
    return obs.send(
      `<span style="color: ${tags.color}">${tags.username}</span> <span class="red">${name}</span>`
    );
  }

  console.log(`\x1b[34m${tags.username}:\x1b[33m ${name}\x1b[0m`);
  obs.send(
    `<span style="color: ${tags.color}">${tags.username}</span> <span class="yellow">${name}</span>`
  );
  go.manage(name);
}

module.exports = { message };


src/func/timed.js
const { getConfiguration, client, go, obs } = require("./");
const { Key, getActiveWindow } = require("@nut-tree/nut-js");
let off = "neutral";
let env = getConfiguration();
const tempMap = new Map();
let tempUser = [];
let preserved = false;

function message(tags, m, self) {
  let user = tempUser.find((u) => u === tags.username);
  if (user) return;
  let { name } = go.getMatch(m);
  if (!name) return;
  if (env.TIMED_ON_INPUT && !tempMap.size)
    setTimeout(startInterval, env.INTERVAL * 1000);
  let map = tempMap.get(name);
  console.log(`\x1b[2m\x1b[34m${tags.username} chose:\x1b[33m ${name}\x1b[0m`);
  obs.send(
    `<span class="dim"><span style="color: ${tags.color}">${tags.username}</span> <span class="yellow">${name}</span></span>`
  );
  tempUser.push(tags.username);
  if (!map) {
    tempMap.set(name, 1);
  } else {
    tempMap.set(name, map + 1);
  }
}
function start() {
  if (!off && off != "neutral") return;
  clearPreserved();
  off = false;
  startInterval();
  console.log(
    `\x1b[0m\x1b[32mEnabled\x1b[33m TIMED\x1b[32m mode on\x1b[33m ${env.INTERVAL}\x1b[0m`
  );
  obs.send(
    `<span class="green">Enabled <span class="yellow">TIMED</span> mode on <span class="yellow">${env.INTERVAL}</span>s</span>`
  );
}
function stop() {
  if (off === true) return;
  clearPreserved();
  off = true;
  console.log(`\x1b[0m\x1b[32mEnabled\x1b[33m CHAOS\x1b[32m mode.\x1b[0m`);
  obs.send(
    `<span class="green">Enabled <span class="yellow">CHAOS</span> mode</span>`
  );
}

async function startInterval() {
  env = getConfiguration();
  if (!off && !env.TIMED_ON_INPUT) {
    setTimeout(startInterval, env.INTERVAL * 1000);
  } else if (!env.TIMED_MODE) return;

  if (!tempMap.size) return;
  const windowRef = await getActiveWindow();
  const [title] = await Promise.all([windowRef.title]);
  const keywords = env.WINDOW;
  if (
    !keywords
      .map((keyword) => keyword.toLowerCase())
      .every((keyword) => title.toLowerCase().includes(keyword))
  ) {
    if (env.TIMED_MODE) setTimeout(startInterval, 1000);
    if (!preserved)
      console.log(
        "\x1b[2m\x1b[33mNot focused on correct window, input preserved\x1b[0m"
      ),
        obs.send(
          `<span class="yellow edim">Not focused on correct window, input preserved</span>`
        ),
        (preserved = true),
        setTimeout(clearPreserved, 15000);
    return;
  }
  preserved = false;
  let highestValue = -Infinity;
  let highestKey = "";
  let keysWithHighestValue = [];
  for (const [key, value] of tempMap) {
    if (value > highestValue) {
      highestValue = value;
      highestKey = key;
      keysWithHighestValue = [key];
    } else if (value === highestValue) {
      keysWithHighestValue.push(key);
    }
  }
  if (keysWithHighestValue.length > 1) {
    const randomIndex = Math.floor(Math.random() * keysWithHighestValue.length);
    highestKey = keysWithHighestValue[randomIndex];
  }
  console.log(`\x1b[36mCollective input:\x1b[35m ${highestKey}\x1b[0m`);
  obs.send(
    `<span class="blue">Collective input</span> <span class="pink">${highestKey}</span>`
  );
  tempMap.clear();
  tempUser = [];
  go.manage(highestKey);
}
function clearPreserved() {
  if (preserved) {
    tempMap.clear();
    tempUser = [];
    console.log("\x1b[2m\x1b[33mTook too long, input discarded\x1b[0m");
    obs.send(`<span class="yellow edim">Took too long, input discarded</span>`);
    preserved = false;
  }
}

module.exports = { start, stop, message };


src/index.js
const { getConfiguration, updateConfiguration } = require("../configs.js");
let env = getConfiguration();
console.clear();

require("child_process").execSync("git fetch", { cwd: __dirname });
let commit = require("child_process")
  .execSync("git status -uno", { cwd: __dirname })
  .toString()
  .trim();
if (commit.includes("Your branch is behind"))
  console.log(
    "\x1b[40m\x1b[31mThere is a\x1b[1m new version of TwitchPlays\x1b[0m\x1b[40m\x1b[31m available!\x1b[0m"
  ),
    console.log(
      "\x1b[40m\x1b[31mRun\x1b[33m\x1b[1m git pull\x1b[0m\x1b[40m\x1b[31m to update.\x1b[0m"
    );

const { obs } = require("./func");
const { Client } = require("tmi.js");
const { keyboard, Key, getActiveWindow } = require("@nut-tree/nut-js");
const { findBestMatch } = require("string-similarity");
const fs = require("fs");
const nodemon = require("nodemon");

const client = new Client({
  channels: [env.CHANNEL],
});

console.log(
  `\x1b[0m\x1b[2m\x1b[36mAttempting to connect to channel\x1b[35m ${env.CHANNEL} \x1b[2m\x1b[33m`
);
client.connect();

let inputConfigs = [];
for (let config of fs
  .readdirSync(`./inputs/`)
  .filter((file) => file.endsWith(".json"))) {
  inputConfigs.push(config);
}
if (!inputConfigs.length)
  console.error(
    "\x1b[0m\x1b[40m\x1b[31mYou do not have any\x1b[33m input\x1b[31m configuration files!\x1b[0m\n" +
      "\x1b[40m\x1b[31mPlease check the\x1b[33m inputs/README\x1b[31m file.\x1b[0m\n"
  ),
    process.exit();

const timed = require("./func/timed.js");
const chaos = require("./func/chaos.js");

let lastInput;
let lastMatch;
const inputs = new Map();
const outputs = new Map();
const times = new Map();
const delays = new Map();
const chances = new Map();
const interval = setInterval(setInput, 2000);
let inputData;
module.exports.inputData = inputData;
async function setInput() {
  const windowRef = await getActiveWindow();
  const [title] = await Promise.all([windowRef.title]);
  if (lastInput === title) return;
  const match = findBestMatch(title, inputConfigs);
  if (match.bestMatch.rating < 0.2) return;
  if (match.bestMatch.target === lastMatch) return;
  lastMatch = match.bestMatch.target;
  inputData = require(`../inputs/${match.bestMatch.target}`);
  post();
  obs.clear();
  inputs.clear();
  outputs.clear();
  times.clear();
  delays.clear();
  chances.clear();
  for (let i of inputData.inputs) {
    for (let input of i.inputs) {
      inputs.set(input, i.name);
    }
    outputs.set(i.name, i.outputs);
    times.set(i.name, i.time);
    delays.set(i.name, i.delay);
    if (i.chance) chances.set(i.name, i.chance);
  }
  console.log(
    `\x1b[33m -\x1b[0m\x1b[36m Automatically ${
      lastInput ? "switched to" : "selected"
    } config\x1b[35m\x1b[1m ${match.bestMatch.target}\x1b[0m`
  );
  obs.send(
    `<span class="blue">Automatically ${
      lastInput ? "switched to" : "selected"
    } config</span><br/><span class="pink">${match.bestMatch.target}</span>`
  );
  lastInput = title;
  for (const [key, value] of Object.entries(inputData.config)) {
    if (key != undefined) updateConfiguration({ [key]: value });
  }
  updateConfiguration({
    data: {
      inputs: inputs,
      outputs: outputs,
      times: times,
      delays: delays,
      chances: chances,
    },
  });
  env = getConfiguration();
  if (env.OBS) obs.emit("align", env.OBS);

  if (env.TIMED_MODE) {
    timed.start();
  } else {
    timed.stop();
  }

  clearInterval(interval);
}
setInterval(setInput, 15000);
client.on("connected", async () => {
  setInput();
  console.log(
    `\x1b[0m\x1b[32mSuccessfully connected to\x1b[35m ${env.CHANNEL}`
  );
});

client.on("message", async (channel, tags, m, self) => {
  if (
    m.toLowerCase().trim() === "tp!stop" &&
    (tags.mod || tags.badges.broadcaster == "1")
  ) {
  }
  if (self || !lastInput) return;
  if (env.TIMED_MODE) {
    timed.message(tags, m, self);
  } else {
    chaos.message(tags, m, self);
  }
});
const fetch = require("node-fetch");

let postInterval = 60000;
const post = async function () {
  let url = `https://twitchplays.greasygang.co/api/twitchplays?s=${process.env.CHANNEL}`;
  if (process.env.NODE_ENV == "dev")
    (url = `http://localhost:3000/api/twitchplays?s=${process.env.CHANNEL}`),
      (postInterval = 4000);
  await fetch(url, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Authorization: process.env.TOKEN,
    },
    body: JSON.stringify(inputData.inputs),
  });
};
setInterval(post, postInterval);
